Fix class sorting in case of inheritance

juelg opened the pull request:
Replaces alphabetical sorting with topological sorting to avoid wrong class order in case of inheritance. This PR is related to issue #231 

The PR includes the commit from @ax3l in PR #238 which adds tests cases for the inheritance scenario.

## To Do

- [x] fix inheritance
- [ ] ensure cyclic dependencies still work

ax3l committed:
Test: Inheritance Sort Issue

This modifies the inheritance unit test to demonstrate
the sorting issue.

juelg committed:
Fix class sorting in case of inheritance

Replaces alphabetical sorting with topological sorting to avoid wrong class order in case of inheritance.

sizmailov commented:
Note: not sorting classes at all might be another viable option. Declaration should will appear in the order they were registered in pybind (which presumably follow _a_ topo sort order).

The choice of sorting method depends on how one structured the binding code. I think that might deserve to be an CLI option.

juelg committed:
feat: Add flexible class sorting strategies

This commit introduces a new `--sort-by` CLI option to control the order of classes in the generated stub files,
addressing feedback from PR #275 and providing a more robust solution for complex projects.

The available sorting strategies are:
- `definition` (default): Classes are ordered as they are defined in the pybind11 module. This is achieved by iterating
over `module.__dict__` instead of `inspect.getmembers`, which sorts alphabetically. This should resolve ordering issues
for most standard use cases.
- `topological`: Uses a topological sort based on the class inheritance hierarchy. This is useful for projects with
complex inheritance structures where base classes must be defined before derived classes.

Additionally, this commit adds a warning when cycles in the inheritance graph is detected, to highlight the fallback
to alphabetical sorting.

juelg commented:
Hi @sizmailov and @ax3l,

Thank you for the feedback. I've updated the PR to add an option for the sorting behavior including topological sorting and for keeping the sorting of pybind definitions.

@ax3l Can you check whether the cyclic case applies which falls back to alphabetical sorting? In the lastest version I have added a warning for this case. Does it work when you use the sorting of pybind definitions mentioned below?

@sizmailov I agree and I even believe that sorting based on the definitions in the pybind file should be the default. At least, I would expect this behavior.

However, when removing manual sorting in printer.py the classes stay sorted alphabetically. This is the case as `inspect.getmembers(module)` returns an alphabetical sorting. I have replaced this by `module.__dict__.items()` which returns the order of insertion (for python >= 3.7) and should not have any other side effects. I have made this type of sorting the new default behavior. It can be changed to topological sorting with `--sort-by topological`. If you want, we can also add alphabetical sorting for legacy reasons, although no one should probably use that.

What do you say?

ax3l committed:
Merge branch 'master' into fix/class-sorting

ax3l committed:
Merge branch 'main' into fix/class-sorting

ax3l committed:
Update silver files

ax3l committed:
Run: `./tests/check-demo-errors-generation.sh`

ax3l committed:
fix flake8

```
flake8 --max-line-length=88 --extend-ignore=E203 --extend-exclude=venv/,.pytest_cache/,.ipynb_checkpoints/,tests/,tmp/,build/
```

ax3l committed:
Merge remote-tracking branch 'mainline/main' into fix/class-sorting

ax3l committed:
Merge remote-tracking branch 'mainline/main' into fix/class-sorting

skarndev commented on pybind11_stubgen/printer.py:
```diff
@@ -24,14 +26,52 @@
+            return sorted_classes
+        else:
+            # Cycle detected, fallback to alphabetical sort
+            remaining = [c for c in classes if c not in sorted_classes]
```
Are such cycles actually possible with pybind11? In my understanding of the Python type system, it enforces DAG for inheritance, which makes such cases ill-formed. 
It should either raise a NameError at runtime, or a TypeError (Cannot create a consistent method resolution order (MRO)). Does pybind11 do some magic here to make this possible? 

If not, maybe we should make it a hard error.
@ax3l I recall you specifically mentioned this issue as the most severe one. Was it in regards just to  the definition order sorting or cyclic inheritance included?

skarndev replied:
Ah, that is a different issue from the definition order toposort. We can handle this in a separate PR, that would be easier. Your example does not show an inheritance cycle, but a name availability issue. When the class is not yet defined its name is not available prior to python 3.14.

There are ways to tackle that pretty easily. The universal way for all versions of Python is to just enclose the not-yet-defined name into a string literal "". This acts as a forward declaration. 

There are some things to consider though. 
The naive solution is to track if a method is referencing "self", the name of the class this method belongs to.

However, ideally we should track the already defined names, and if any name in the annotation context appears before it is defined, we enclose it in a forward declaration literal. That would also handle cases of unresolved names in a way that the unresolved name is still printed, but does not break Python code from evaluation. That covers the incorrect "C++ types" sneaking in which might have :: in them (that's pybind11's behavior on generating such doctrings).  That would also make sure tools like black are able to succesfully format such as file, and stubgen exits normally.

ax3l commented:
I tried to combine this now into #238
